File: languages/csharp/02-core/sorting-and-searching/example/stable-vs-unstable-sort.cs
// This helper example focuses on contrasting two sorting behaviors without extra setup noise.

// This extra example extends sorting-and-searching with tie-order comparisons.

using System;
using System.Collections.Generic;
using System.Linq;

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
class Record
{
    public int Score { get; init; }

    public string Name { get; init; } = string.Empty;
}

class Program
{
    static void PrintRecords(IEnumerable<Record> records, string title)
    {
        Console.WriteLine(title);
        foreach (Record record in records)
        {
            Console.WriteLine($"- score={record.Score}, name={record.Name}");
        }
        Console.WriteLine();
    }

    static void Main()
    {
        List<Record> records = new List<Record>
        {
            new Record { Score = 90, Name = "Ana" },
            new Record { Score = 75, Name = "Bob" },
            new Record { Score = 90, Name = "Carla" },
            new Record { Score = 75, Name = "Diego" },
            new Record { Score = 90, Name = "Emma" },
        };

        PrintRecords(records, "Original order:");

        List<Record> unstable = new List<Record>(records);
        unstable.Sort((left, right) => left.Score.CompareTo(right.Score));
        PrintRecords(unstable, "After List.Sort (not guaranteed stable):");

        List<Record> stable = records.OrderBy(record => record.Score).ToList();
        PrintRecords(stable, "After LINQ OrderBy (stable):");

        Console.WriteLine("Notice how equal scores keep relative order with OrderBy.");
    }
}

File: languages/python/02-core/sorting-and-searching/example/stable-vs-unstable-sort.py
# This helper example focuses on contrasting two sorting behaviors without extra setup noise.

# This extra example extends sorting-and-searching with tie-order comparisons.

# Keep this helper separate so the main example can focus on the larger idea without extra noise.
def print_records(records, title):
    print(title)
    for record in records:
        print(f"- score={record['score']}, name={record['name']}")
    print()


records = [
    {"score": 90, "name": "Ana"},
    {"score": 75, "name": "Bob"},
    {"score": 90, "name": "Carla"},
    {"score": 75, "name": "Diego"},
    {"score": 90, "name": "Emma"},
]

print_records(records, "Original order:")

stable = sorted(records, key=lambda record: record["score"])
print_records(stable, "After sorted(...) by score (stable):")

reordered_ties = sorted(records, key=lambda record: (record["score"], record["name"]))
print_records(reordered_ties, "After sorted(...) with explicit tie-breaker:")

print("Python's sort is stable. Tie-breakers can still change the order of equal scores.")
